File: go/105/main.go
package main

import "fmt"

func main() {
	pre := []int{3, 9, 20, 15, 7}
	in := []int{9, 3, 15, 20, 7}

	root := buildTree(pre, in)
	printTree(root)
}

func printTree(root *TreeNode) {
	if root == nil {
		fmt.Print("nil ")
		return
	}

	fmt.Printf("%v ", root.Val)
	printTree(root.Left)
	printTree(root.Right)
}

/**
 * Definition for a binary tree node.
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree(preorder []int, inorder []int) *TreeNode {
	n := len(preorder)
	return buildSubtree(preorder, inorder, 0, n-1, 0, n-1)
}

func buildSubtree(pre []int, in []int, preStart, preEnd, inStart, inEnd int) *TreeNode {
	// Special case: start > end
	if preStart > preEnd || inStart > inEnd {
		return nil
	}

	// Construct the root node
	rootVal := pre[preStart]
	root := &TreeNode{
		Val:   rootVal,
		Left:  nil,
		Right: nil,
	}

	// Special case: only one element is in range
	if preStart == preEnd {
		return root
	}

	// Find the idx of the root in in and split pre and in by the idx
	idxInRoot := -1
	for i := inStart; i <= inEnd; i++ {
		if in[i] == rootVal {
			idxInRoot = i
			break
		}
	}

	// Panic if the root value is not found in in
	if idxInRoot == -1 {
		panic("Invalid input")
	}

	lenLeft := idxInRoot - inStart

	root.Left = buildSubtree(pre, in, preStart+1, preStart+lenLeft, inStart, idxInRoot-1)
	root.Right = buildSubtree(pre, in, preStart+lenLeft+1, preEnd, idxInRoot+1, inEnd)

	return root
}

File: java/105/Main.java
import java.util.ArrayDeque;
import java.util.Optional;
import java.util.Queue;

public class Main {
	public static void main(final String[] args) {
		final var s = new Solution();

		final var p1 = new int[] { 3, 9, 20, 15, 7 };
		final var i1 = new int[] { 9, 3, 15, 20, 7 };

		// Expecting [3, 9, 20, null, null, 15, 7]
		printTree(s.buildTree(p1, i1));
	}

	private static void printTree(final TreeNode root) {
		if (root == null) {
			System.out.println("[]");
			return;
		}

		System.out.print("[");
		var isFirst = true;

		final Queue<Optional<TreeNode>> q = new ArrayDeque<>();
		q.offer(Optional.of(root));

		while (!q.isEmpty()) {
			final var layerLen = q.size();
			var isNextLayerAllEmpty = true;

			for (var i = 0; i < layerLen; i++) {
				if (!isFirst) {
					System.out.print(", ");
				}

				final var node = q.poll();
				if (node.isEmpty()) {
					System.out.print("null");
				} else {
					final var unwrappedNode = node.get();
					System.out.print(unwrappedNode.val);
					q.offer(Optional.ofNullable(unwrappedNode.left));
					q.offer(Optional.ofNullable(unwrappedNode.right));
					if (isNextLayerAllEmpty && (unwrappedNode.left != null || unwrappedNode.right != null)) {
						isNextLayerAllEmpty = false;
					}
				}

				isFirst = false;
			}

			if (isNextLayerAllEmpty) {
				break;
			}
		}

		System.out.println("]");
	}
}

/**
 * Definition for a binary tree node.
 */
class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;

	TreeNode() {
	}

	TreeNode(final int val) {
		this.val = val;
	}

	TreeNode(final int val, final TreeNode left, final TreeNode right) {
		this.val = val;
		this.left = left;
		this.right = right;
	}
}

class Solution {
	public TreeNode buildTree(final int[] preorder, final int[] inorder) {
		final var root = getRoot(preorder, inorder, 0, preorder.length - 1, 0, inorder.length - 1);
		return root;
	}

	// preorderRight and inorderRight are both inclusive
	private TreeNode getRoot(final int[] preorder, final int[] inorder, final int preorderLeft, final int preorderRight, final int inorderLeft,
			final int inorderRight) {
		if (inorderLeft > inorderRight) {
			return null;
		}

		final var rootVal = preorder[preorderLeft];
		var inorderSplitIdx = inorderLeft;
		while (inorderSplitIdx <= inorderRight) {
			if (inorder[inorderSplitIdx] == rootVal) {
				break;
			}

			inorderSplitIdx++;
		}

		final var leftCount = inorderSplitIdx - inorderLeft;
		final var root = new TreeNode(rootVal,
				getRoot(preorder, inorder, preorderLeft + 1, preorderLeft + leftCount, inorderLeft,
						inorderSplitIdx - 1),
				getRoot(preorder, inorder, preorderLeft + leftCount + 1, preorderRight, inorderSplitIdx + 1,
						inorderRight));
		return root;
	}
}
